Global styles: Allow indirect properties when unfiltered_html is not allowed

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #45520

This PR adds a new `INDIRECT_PROPERTIES_METADATA` array to the theme JSON class, to store properties metadata for properties that are not directly output by `compute_style_properties`, but are instead used elsewhere in global styles output.

This array is then used in `remove_insecure_styles` to ensure that style values that will be used in other cases are validated against an indirect property. The main use case at the moment is for `blockGap` style values, which are output as either top margins or gap styles, depending on the layout type. For the purposes of validation, though, we can safely validate against a small list of gap properties.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Over in #45520 it was flagged that the absence of `blockGap` in `PROPERTIES_METADATA` meant that the value was being stripped in situations where a user doesn't have the `unfiltered_html` capability and is saving global styles (e.g. for multisite regular admins). Hopefully, this PR should resolve that.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add an `INDIRECT_PROPERTIES_METADATA` array to the theme JSON class.
* Validate against this array in `remove_insecure_styles`.
* Add a Gutenberg version of `wp_filter_global_styles_post` (`gutenberg_filter_global_styles_post`) so that we can test this change in Gutenberg
* If the core versions of `wp_filter_global_styles_post` callbacks are registered, then de-register, and re-register the Gutenberg versions instead

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

Note: a quick way to test this PR is to place the following in `lib/init.php` in the Gutenberg plugin, rather than worrying about changing user capabilities:

```
add_action( 'init', 'kses_init_filters' );
```

1. With this PR applied and with a user with `unfiltered_html` capability (e.g. a regular admin), you should be able to save root block spacing values in global styles, as well as block level spacing (e.g. set Buttons spacing, and axial row/column spacing for Social Icons).
2. Repeating the above with a user _without_ `unfiltered_html` capability (or enable the KSES filters with `add_action( 'init', 'kses_init_filters' );`) you should now still be able to save root block spacing values in global styles, and block level spacing.

Ensure tests pass:

```
npm run test:unit:php -- --filter WP_Theme_JSON_Gutenberg_Test
```

## Screenshots or screencast <!-- if applicable -->

| Before (note value is reset when saved) | After (value is saved) |
| --- | --- |
| ![2022-12-08 11 32 33](https://user-images.githubusercontent.com/14988353/206327161-2e80181f-4a4f-4786-95a5-afe4603e7a2c.gif) | ![2022-12-08 11 28 44](https://user-images.githubusercontent.com/14988353/206326699-bd2f9cc5-1fd7-4bfa-b3eb-fd66f6c5c4d9.gif) |

andrewserong committed:
Global styles: Allow inferred properties outside of compute_style_properties when unfiltered_html is not allowed

andrewserong committed:
Ensure set_current_user callback is covered

andrewserong committed:
Add a simple test

mmtr commented on lib/experimental/class-wp-theme-json-gutenberg.php:
```diff
@@ -29,3 +29,56 @@ class WP_Theme_JSON_Gutenberg extends WP_Theme_JSON_6_2 {
+ * This function should not be backported to core.
+ */
+function gutenberg_override_core_kses_init_filters() {
+	if ( has_filter( 'content_save_pre', 'wp_filter_global_styles_post' ) ) {
+		remove_filter( 'content_save_pre', 'wp_filter_global_styles_post', 9 );
+		add_filter( 'content_save_pre', 'gutenberg_filter_global_styles_post', 9 );
+	}
+
+	if ( has_filter( 'content_filtered_save_pre', 'wp_filter_global_styles_post' ) ) {
+		remove_filter( 'content_filtered_save_pre', 'wp_filter_global_styles_post', 9 );
+		add_filter( 'content_filtered_save_pre', 'gutenberg_filter_global_styles_post', 9 );
+	}
```
Nitpick (non-blocking): `remove_filter` returns true if the given callback existed before it was removed, so this can be slightly simplified.
```suggestion
	if ( remove_filter( 'content_save_pre', 'wp_filter_global_styles_post', 9 ) ) {
		add_filter( 'content_save_pre', 'gutenberg_filter_global_styles_post', 9 );
	}

	if ( remove_filter( 'content_filtered_save_pre', 'wp_filter_global_styles_post', 9 ) ) {
		add_filter( 'content_filtered_save_pre', 'gutenberg_filter_global_styles_post', 9 );
	}
```

Although using `has_filter` explicitly might be clearer and avoid potential confusion, so feel free to keep this as-is!

andrewserong replied:
Another good question — in this case, I personally prefer the explicit `has_filter` check so that it's clearer at a glance that the Gutenberg filter is only added if a corresponding core one existed. This deviates a little from how we typically register Gutenberg filters in that we usually expect them to always be running, but in this case, it should only be registered for the use case of no `unfiltered_html` capability.

Again, happy to use the simplified approach if folks prefer! 🙂

mmtr commented on lib/compat/wordpress-6.2/class-wp-theme-json-6-2.php:
```diff
@@ -216,4 +235,45 @@ class WP_Theme_JSON_6_2 extends WP_Theme_JSON_6_1 {
+			}
+		}
+
+		foreach ( static::INFERRED_PROPERTIES_METADATA as $property => $path ) {
+			$value = _wp_array_get( $input, $path, array() );
+			if (
+				isset( $value ) &&
+				! is_array( $value ) &&
+				static::is_safe_css_declaration( $property, $value )
+			) {
+				_wp_array_set( $output, $path, $value );
+			}
+		}
```
Is there any reason why this shouldn't be done in `remove_insecure_properties` instead? I see that method has a similar logic to [bring back pseudo selectors that are removed by `remove_insecure_styles`](https://github.com/WordPress/gutenberg/blob/af8bf6e305dc8f2478c8991b8eb2dd92b4b6456e/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php#L316-L325), so maybe it makes sense to have the new logic there too for consistency?

I guess the implementation would be slightly harder that way. And maybe this is the right place since this logic is strictly related to how `compute_style_properties` behaves.

Feel free to disregard this question if it's silly 😅

andrewserong replied:
Not at all a silly question! My thinking behind this one was that the logic surrounding when to include inferred properties directly relates to the check for declarations based on the current `$input`, so is more closely related to the logic inside of `remove_insecure_styles`, rather than the parent function. Over in `remove_insecure_properties`, the logic that re-adds pseudo selectors needs to call `remove_insecure_styles` again using the existing logic, so I was imagining this particular logic change as acting more closely to the leaf in the tree, if that makes sense.

Happy to look into changing it if folks prefer, though!

mmtr replied:
That makes sense, thanks!

mmtr reviewed (commented):
Lovely, thanks for working on this!

It tests quite well for me, and the proposed approach makes sense to me. I'd love though to hear more feedback from someone with more experience than me in this area such as @ramonjd or @tellthemachines, since I'm not confident enough to approve this.

I left a couple of minor comments, but to me this is ready to go.

andrewserong commented:
Thanks for giving this a test and for adding your thoughts @mmtr! 🙇

> I'd love though to hear more feedback from someone with more experience than me in this area

Good thinking! In that case, I'll just ping @WordPress/gutenberg-core and @scruffian too for visibility to see if there's anyone else with thoughts on the direction to take here, or potential implications on changing the logic here to add the concept of inferred properties in addition to explicit/direct properties that are handled by `compute_style_properties`.

tellthemachines commented on lib/compat/wordpress-6.2/class-wp-theme-json-6-2.php:
```diff
@@ -97,6 +97,25 @@ class WP_Theme_JSON_6_2 extends WP_Theme_JSON_6_1 {
 	);
 
+	/**
+	 * Inferred metadata for style properties that are not directly output.
```
I'm not sure "inferred" quite translates the status of these properties, could it instead be something like "indirect"? Or perhaps even "auxiliary" as we use them as helpers for layout.

andrewserong replied:
Oh, I quite like "indirect", that gets closer to the meaning I was going for than "inferred". I'll give that a try 👍

ramonjd replied:
> could it instead be something like "indirect"?

➕ 

Much better than what I was going to suggest: `APPLIED_PROPERTIES_METADATA` or `EVENTUALLY_APPLIED_PROPERTIES_METADATA` :trollface:

tellthemachines commented on lib/experimental/class-wp-theme-json-gutenberg.php:
```diff
@@ -29,3 +29,56 @@ class WP_Theme_JSON_Gutenberg extends WP_Theme_JSON_6_2 {
+
+}
+add_action( 'init', 'gutenberg_override_core_kses_init_filters' );
+add_action( 'set_current_user', 'gutenberg_override_core_kses_init_filters' );
```
Should these changes be in a `kses.php` file instead, to mirror core? Or perhaps we should have a file somewhere for changes that are not ever meant to be backported? It's best to keep class files for just the class.

andrewserong replied:
> Should these changes be in a kses.php file instead, to mirror core?

Great point, yes! The `lib/experimental` directory appears to be a good place for things that won't be backported necessarily since it isn't yet in a release, so I think `lib/experimental/kses.php` might work. I can give that a go 🙂

tellthemachines reviewed (commented):
Testing locally with the `add_action( 'init', 'kses_init_filters' )` snippet, this fixes the issue ✅ 

Just a couple of readability/code organisation points below, otherwise code looks good!

ramonjd commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -751,6 +751,30 @@ public function test_get_stylesheet_with_block_support_feature_level_selectors()
+			),
+		);
+
+		$this->assertEqualSetsWithIndex( $expected, $actual );
```
Would we need to cover block-level spacing as well or is that being piped through the same mechanism?

andrewserong replied:
Turns out I think I made a typo when I tried the block-level spacing in this test earlier, but it's working now that I've written it out correctly 😄. Added in https://github.com/WordPress/gutenberg/pull/46388/commits/5b629c87a04d2c18171b690348d7515e31b04119.

ramonjd reviewed (approved):
Looking pretty good! Thanks for working on this.

On a multisite installation I created an admin and tried to set blockGap:


| Trunk | This branch |
| ------------- | ------------- |
| ![2022-12-12 11 01 34](https://user-images.githubusercontent.com/6458278/206938700-1fe896ee-cf02-46c2-809d-31d9f9c2a92b.gif) | ![2022-12-12 11 21 59](https://user-images.githubusercontent.com/6458278/206938707-6880027f-71f2-4657-85ea-19c46a0336e4.gif)  |

Had a couple of non-blocking questions around naming and tests.

Do we need a dev note for this? I'm thinking yes since it's affecting WP 6.1?

andrewserong commented:
Thanks for the reviews @tellthemachines and @ramonjd! I'll rename the property and move the code around a little bit. 

> Would we need to cover block-level spacing as well or is that being piped through the same mechanism?

It winds up being piped through the same mechanism. The test currently just looks at the root level — I _think_ to get the block level working in the test, we might need to register blocks, too 🤔. The change in this PR doesn't explicitly look at the nesting / recursive behaviour of `remove_insecure_properties` so I wasn't sure how much of a deal breaker it might be to not test it explicitly, but happy to dig in a bit further!

> Do we need a dev note for this? I'm thinking yes since it's affecting WP 6.1?

Good call. Once this PR lands, I'll open up a corresponding backport PR for core. I think an appropriate time for a dev note could be once the backport PR is approved, in case there's any additional feedback / work required for the core side of things. Let me know if you think a dev note should be put up sooner, though!

ramonjd commented:
>  I think an appropriate time for a dev note could be once the backport PR is approved, in case there's any additional feedback / work required for the core side of things. Let me know if you think a dev note should be put up sooner, though!

Sounds good. Cheers!

andrewserong committed:
Renamed to indirect, move hooks to separate kses file

andrewserong committed:
Add block-level styles to the test

andrewserong commented:
Update: I've done the following:

* Renamed `inferred` to `indirect`
* Moved hooks to a new `kses.php` file
* Updated the test to also incorporate block-level spacing and make sure it isn't stripped

I'll leave this PR open overnight just in case anyone else pinged would like to chime in, but otherwise I think we should be able to merge this in tomorrow 🤞

tellthemachines reviewed (approved):
Changes look good!

andrewserong commented:
Thanks again for the reviews! Since this PR only updates internal implementation details of the theme JSON class, I'll merge this in now. If there's any follow-up feedback, or if folks would like further changes, we should be able to happily make changes without affecting the API at all.

andrewserong commented:
## Dev note for 6.2

This bug fix is an internal implementation detail of the Theme JSON class, and doesn't _really_ directly affect extenders / people working with `theme.json` in that no changes are required for themes. So, it probably fits best in a Miscellaneous dev note. Draft below:

---

In WordPress 6.1 the [layout block support was refactored](https://make.wordpress.org/core/2022/10/10/updated-editor-layout-support-in-6-1-after-refactor/) to output CSS values for block spacing, with additional layout controls within global styles being made available. This resulted in a bug for admin users without the `unfiltered_html` capability (e.g. admin users within a multisite instance) where block spacing or layout content sizes were not saved when updated within the global styles interface. The cause was that within the Theme JSON class, `theme.json` values are filtered for users without the `unfiltered_html` capability based on a lookup to a direct CSS property, e.g. `margin` or `padding`. For layout features, the styling rules are stored in _indirect_ properties that don't directly map to a real CSS property, e.g. `blockGap` and `contentSize`.

For WordPress 6.2 (and backported to WordPress 6.1.2) this issue is resolved by the Theme JSON class storing a list of indirect properties (`INDIRECT_PROPERTIES_METADATA`), with a mapping between a CSS property to use for validation, and the path to the indirect property as stored in `theme.json`. In this way, `blockGap` will be validated against the CSS `gap` property, and `contentSize` against the CSS `max-width` property, etc.

For themes using layout features, no changes are required, as the bug fix is an internal implementation detail of how Theme JSON styles are validated and output.

### Further reading

* https://github.com/WordPress/gutenberg/pull/46388
* https://github.com/WordPress/gutenberg/pull/46712
* Reported issue: https://github.com/WordPress/gutenberg/issues/45520
* Core fix: https://github.com/WordPress/wordpress-develop/pull/3742
* Trac ticket: https://core.trac.wordpress.org/ticket/57321

---

@bph just sharing a draft dev note above — this can be combined with other misc changes in 6.2, I think.
